### PR TITLE
Fix path used in sabre's tree cache

### PR DESCRIPTION
**Pull request by icewind1991**
Sabre's `getChildren()` cached path with a leading slash, now we properly used entries cached by `getChildren()` which prevents us from having to re-query the fileinfo.

Greatly improves performance when doing a Depth=1 PROPFIND ([comparison](https://blackfire.io/profiles/compare/e42556de-fc7a-4e16-8588-0b7284c91541/graph))

cc @DeepDiver1975 @PVince81 @LukasReschke

### CODE REVIEW

**Comment by DeepDiver1975** [comment]
is there unit test coverage on the tree?

**Comment by icewind1991** [comment]
Yes

**Comment by rullzer** [comment]
:-1: 

Let me explain. What happens is that the cache is used and filled in the function `getNodeForPath`. But it is also filled in the function `getChildren` https://github.com/owncloud/3rdparty/blob/master/sabre/dav/lib/DAV/Tree.php#L192-L204

The trouble is at https://github.com/owncloud/3rdparty/blob/master/sabre/dav/lib/DAV/Tree.php#L196 that for child nodes puts a '/' between the `basePath` and the childs name. Now of course when you are in the root there is no `basePath` so the child nodes get cached like `/<CHILD>`. However if you a subfolder they get cached like `<SUBFOLDER>/<CHILD>`.

So your patch will speedup PROPFINDS in the root but slow down in all subfolders.

Basically this is a bug in sabre. Or at least in how we use it ;) We could fix it by copyng the code of `getChildren` over an fixing it. But probably better to fix upstream. @evert do you want me to prepare a PR against 2.1.7 or will there be no 2.1.8?

**Comment by rullzer** [comment]
Upstream PR in https://github.com/fruux/sabre-dav/pull/729

**Comment by rullzer** [comment]
Of course I did some measurements:

propfind on root with 5 folders and 5 files

what | time | graph
---- | ---- | -----
master | 306ms | [graph](https://blackfire.io/profiles/ddf0e52e-cdeb-47e1-a31d-4ea6575cc520/graph)
master + this patch | 70.7ms | [graph](https://blackfire.io/profiles/c63397b6-7c66-43c2-993d-f589e9e7000f/graph)
master + sabre patch | 47.3ms | [graph](https://blackfire.io/profiles/46497ea3-503b-404b-b32e-8a8c75c89596/graph)

propfind on subfolder with 5 folders and 5 files

what | time | graph
---- | ---- | -----
master | 72.9ms | [graph](https://blackfire.io/profiles/afd3f1cc-e0c4-4e51-83c0-bb2d102fd9a0/graph)
maser + this patch | 223ms | [graph](https://blackfire.io/profiles/38db35b1-2cee-41af-beff-f9ae64ea74e3/graph)
master + sabre patch | 69.6ms | [graph](https://blackfire.io/profiles/acd187d8-69a3-4be2-b5c5-d656459fd1f7/graph)

**Comment by LukasReschke** [comment]
Awesome work @icewind1991 and @rullzer :rocket:

**Comment by PVince81** [comment]
I suggest to provide a method "setCachedEntry" that auto-adds or auto-strips the leading slash. It's not the first time this kind of thing happens, had this lately with Dropbox local cache too: see https://github.com/owncloud/core/pull/19877

**Comment by rullzer** [comment]
Fix is in sabre master. So we have yet another reason to upgrade ;)
Let close this then.

**Comment by DeepDiver1975** [comment]
> Fix is in sabre master. So we have yet another reason to upgrade ;)
> Let close this then.

in order to get this fixed in 8.2 and earlier we need a fix in 2.1 - requested at https://github.com/fruux/sabre-dav/commit/6223d06b7e0e87e0253542db31d12bd8ec967c96#commitcomment-14005068

**Comment by RealRancor** [comment]
@rullzer @DeepDiver1975 @icewind1991 Just stumbled over this. 2.1.9 (https://github.com/owncloud/core/issues/21914) has included this fix.
